src/gt4py/next/common.py

```diff
@@ -117,10 +129,102 @@ def __getitem__(self, index: int | slice) -> int | UnitRange:
             else:
                 raise IndexError("UnitRange index out of range")
 
+    def __and__(self, other: UnitRange) -> UnitRange:
+        start = max(self.start, other.start)
+        stop = min(self.stop, other.stop)
+        return UnitRange(start, stop)
+
 
 DomainT: TypeAlias = tuple[tuple[Dimension, UnitRange], ...]
 
 
+from dataclasses import dataclass
+from collections.abc import Sequence
+
+
+@dataclass(frozen=True)
+class Domain(Sequence):
+    dims: list[Dimension]
+    ranges: list[UnitRange]
+
+    def __len__(self) -> int:
+        return len(self.ranges)
+
+    def __iter__(self):
+        return iter(zip(self.dims, self.ranges))
+
+    def __contains__(self, item: object) -> bool:
+        if isinstance(item, int):
+            return any(item in range_ for range_ in self.ranges)
+        elif isinstance(item, UnitRange):
+            return any(
+                item.start >= range_.start and item.stop <= range_.stop for range_ in self.ranges
+            )
+        return False
+
+    @overload
+    def __getitem__(self, index: int) -> tuple[Dimension, UnitRange]:
+        ...
+
+    @overload
+    def __getitem__(self, index: slice) -> "Domain":
+        ...
+
+    @overload
+    def __getitem__(self, index: Dimension) -> tuple[Dimension, UnitRange] | "Domain":
+        ...
+
+    def __getitem__(self, index):
+        if isinstance(index, int):
+            return self.dims[index], self.ranges[index]
+        elif isinstance(index, slice):
+            start, stop, step = index.indices(len(self))
+            dims_slice = self.dims[start:stop:step]
+            ranges_slice = self.ranges[start:stop:step]
+            return Domain(dims_slice, ranges_slice)
+        elif isinstance(index, Dimension):
+            dim_range_subset = [
+                (dim, range_) for dim, range_ in zip(self.dims, self.ranges) if dim == index
+            ]
+            if len(dim_range_subset) > 1:
+                return Domain(
+                    [dim_range[0] for dim_range in dim_range_subset],
+                    [dim_range[1] for dim_range in dim_range_subset],
+                )
+            elif len(dim_range_subset) == 1:
```

Comment:
just assert that Dimension is unique for now

Reply:
I've added a `__post_init__` check for this which simplifies `__get_item__`